### ISSUE TITLE
Transactions recover state data bug (GrainReferenceNotBoundException)

### ISSUE THREAD

**Issue by 453873** [dotnet/orleans#8589]
1. Multiple entities participate in a transaction.
2. Sudden interruption of the transaction (shutdown program) during the write transaction
3. Restart the program, execute a new transaction, and one of the data participating in the transaction contains the data that failed the previous transaction
4. GrainReferenceNotBoundException will be reported during transaction execution

In the Load() method of the implemented ITransactionalStateStorage<T> interface code, it is found that TransactionManager.Reference.Runtime in pendingStates is null.(CommitRole is RemoteCommit)

As a temporary solution, in the Load() method of the ITransactionalStateStorage<T> interface code, set the Runtime for calling GrainFactory.BindGrainReference(transactionManager.Reference) for pendingStates.

**Comment by HermesNew** [comment]
I have also encountered similar issues, such as database restart, which can cause previously incomplete transactions to not load correctly. After debug, it was found that GrainReference.Runtime is null.
error logs:
![image](https://github.com/dotnet/orleans/assets/22929457/e8151ce2-d161-4dab-816e-d190b227c628)

![0RN ~%N6X2EP)%_2T1{{9`O](https://github.com/dotnet/orleans/assets/22929457/36b426da-b2b3-4de1-be65-c449d313c0ac)

If Runtime==null, I think the binding should be re executed to fix this issue.
![~9 CQSR IFN8W ZR9VO62`U](https://github.com/dotnet/orleans/assets/22929457/4f26a67e-4b9c-40c6-b4ed-0c06ac1130c9)

This property can be used to determine whether it is necessary to rebind the runtime.

**Comment by HermesNew** [comment]
BTW Orleans version is 3.7.1

**Comment by HermesNew** [comment]
@ReubenBond This is a reliability bug ，that needs to be fixed.

**Comment by ReubenBond** [comment]
Agreed. Are you able to create a PR? Otherwise we will try to get to it soon.
